### ISSUE TITLE
Improve fingerprinting mechanism on checkout page

### ISSUE THREAD

**Issue by tpaksu** [Automattic/woocommerce-payments#4772]
### Description
We need to implement a more intelligent fingerprint solution that doesn’t only rely on HTTP headers, and probably we can implement a 3rd party solution like fingerprint.js, or Sift’s fingerprint library, or another one, but there’s a problem that their scripts can be blocked by ad-blockers. We need to find a way to safely load a fingerprinting library which can’t be blocked. To be sure, we can block loading our scripts if the fingerprint library is not loaded, but we need to check first if it’s feasible.

### Dev notes
Some 3rd party fingerprinting scripts are blocked automatically by ad blockers, you can get some insight of what's blocked from [this](https://github.com/Hackl0us/AdBlock-Rules-Mirror/blob/master/Easylist-Privacy.txt) list.

**Comment by allie500** [comment]
[Fingerprint Solutions Research document in Google Drive](https://docs.google.com/document/d/1o6KL0lvgx7Ue8IdhwN0_4gtt2wH1d9Qt5Y0ZaaueM7I/edit?usp=sharing)

Currently we are looking into leveraging the existing Sift integration.

**Comment by allie500** [comment]
Further investigation revealed that the Sift device fingerprinting API was not being made available to new customers by Sift. We have started looking into using Fingerprint JS. See the [Fingerprint JS Discovery](https://docs.google.com/document/d/1o6KL0lvgx7Ue8IdhwN0_4gtt2wH1d9Qt5Y0ZaaueM7I/edit#heading=h.4rr6vmuyucoa) section of the Google doc.

**Comment by eduardoumpierre** [comment]
I'm currently checking the FingerprintJS custom subdomain setup viability with Systems. pMz3w-ghK-p2

**Comment by eduardoumpierre** [comment]
I'll create a P2 post about the security concerns from the Systems thread so we can discuss and determine what would be the best approach/domain to work with.

A draft PR for the client part [Implement FingerprintJS to the checkout page](https://github.com/Automattic/woocommerce-payments/pull/5071) was created.

The server implementation will be detailed on its own PR and it will be linked here whenever it has been created.

**Comment by tpaksu** [comment]
Hey @eduardoumpierre, thanks for the detailed updates! One question, what do you think of preventing loading the Stripe elements form which is used for UPE, when the fingerprinting script is blocked somehow? I don't know the feasibility on it, and it may lead to bad user experience, but if we have a way to do it seamlessly, it would be perfect. I think you already asked systems for a subdomain, will it be a good measure for avoiding the ad blockers?

**Comment by eduardoumpierre** [comment]
That's a great point @tpaksu. I think the main issue is that ad blockers rely on lists such as [EasyList](https://easylist.to/easylist/easylist.txt), [EasyPrivacy](https://easylist.to/easylist/easyprivacy.txt), etc (more on [uBlock](https://github.com/gorhill/uBlock)), so I'm not sure about the user experience as it could potentially block legit users if ad blockers update their lists.

We could definitely prevent the Stripe elements from loading, but we would need to communicate it clearly to the users and it may impact conversion.

> I think you already asked systems for a subdomain, will it be a good measure for avoiding the ad blockers?

Yes, it should be effective as they are blocking specific URL parts such as `||fptls.com^` and `||fpjs.io^$third-party`. If we set up the subdomain routing and avoid using blocked keywords as `fingerprint`, we should be able to avoid ad blockers.

**Comment by eduardoumpierre** [comment]
As discussed on p1667920761330599/1667917180.793909-slack-C01B8KNUYSW, we are changing our approach for this specific issue.

We are planning to use FingerprintJS (open source) and server-side signals to detect bad actors, which wouldn't require any custom subdomain setup and is less likely to be affected by ad blockers.

The changes will be done directly in the open PR, and this issue will be updated whenever we have the server-side PR open.

**Comment by eduardoumpierre** [comment]
The open-source version of FingerprintJS was implemented on [Implement FingerprintJS on the checkout page](https://github.com/Automattic/woocommerce-payments/pull/5071) for classic and block checkouts. It's currently in draft mode, but the PR's description and testing instructions will be filled in tomorrow so we can code review before the next week.

**Comment by eduardoumpierre** [comment]
Both PRs are ready to be reviewed:
- Client PR - https://github.com/Automattic/woocommerce-payments/pull/5071
- Server PR - https://github.com/Automattic/woocommerce-payments-server/pull/2695
